plans/changes.go

```diff
@@ -277,6 +294,43 @@ func (rc *ResourceInstanceChange) Simplify(destroying bool) *ResourceInstanceCha
 	return rc
 }
 
+// ResourceInstanceChangeActionReason allows for some extra user-facing
+// reasoning for why a particular change action was chosen for a particular
+// resource instance.
+//
+// This only represents sufficient detail to give a suitable explanation to
+// an end-user, and mustn't be used for any real decision-making during the
+// apply step.
+type ResourceInstanceChangeActionReason rune
+
+//go:generate go run golang.org/x/tools/cmd/stringer -type=ResourceInstanceChangeActionReason changes.go
+
+const (
+	// In most cases there's no special reason for choosing a particular
+	// action, which is represented by ResourceInstanceChangeNoReason.
+	ResourceInstanceChangeNoReason ResourceInstanceChangeActionReason = 0
+
+	// ResourceInstanceReplaceBecauseTainted indicates that the resource
+	// instance must be replaced because its existing current object is
+	// marked as "tainted".
+	ResourceInstanceReplaceBecauseTainted ResourceInstanceChangeActionReason = 'T'
+
+	// ResourceInstanceReplaceByRequest indicates that the resource instance
+	// is planned to be replaced because a caller specifically asked for it
+	// to be using ReplaceAddrs. (On the command line, the -replace=...
+	// planning option.)
+	ResourceInstanceReplaceByRequest ResourceInstanceChangeActionReason = 'R'
+
+	// ResourceInstanceReplaceBecauseCannotUpdate indicates that the resource
+	// instance is planned to be replaced because the provider has indicated
+	// that a requested change cannot be applied as an update.
+	//
+	// In this case, the RequiredReplace field will typically be populated on
+	// the ResourceInstanceChange object to give information about specifically
+	// which arguments changed in a non-updatable way.
+	ResourceInstanceReplaceBecauseCannotUpdate ResourceInstanceChangeActionReason = 'F'
```

Comment:
Would we want to make this more vague, like `ResourceInstanceReplaceBecauseProviderSaidSo`, or leave open the possibility that the provider may add other reasons? It may be out of scope for this change, but If we're adding detail here we could make room for new protocol information.

Reply:
I think it's helpful to keep this one specific because as the docs here note we expect this one to come along with a `RequiredReplace` value, and having that additional context to show gives some license to be a little less verbose in the top-level explanation in the plan output because we assume that the `# (forces replacement)` annotations will fill in the gaps. That wouldn't (necessarily) be true for other reasons we might allow a provider to indicate later.

We can always add more of these, of course! If we do add a generalized mechanism later I'd expect something like your `BecauseProviderSaidSo` to appear alongside `BecauseCannotUpdate` to represent a situation where we'd need to match that with a similar explanation in the UI... hopefully something better than "must be replaced because the provider said so", but we'll see what the actual use-case is and see what is a better message.